// @irys/sdk/build/cjs/web/irys.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebIrys = void 0;
const base_1 = require("./base");
const tokens_1 = __importDefault(require("./tokens"));
class WebIrys extends base_1.BaseWebIrys {
    constructor({ url, token, wallet, config, }) {
        super({
            url,
            wallet,
            config,
            getTokenConfig: (irys) => {
                var _a, _b;
                return (0, tokens_1.default)({
                    irys,
                    token: token.toLowerCase(),
                    wallet: (_a = wallet === null || wallet === void 0 ? void 0 : wallet.provider) !== null && _a !== void 0 ? _a : wallet,
                    providerUrl: (_b = config === null || config === void 0 ? void 0 : config.providerUrl) !== null && _b !== void 0 ? _b : wallet === null || wallet === void 0 ? void 0 : wallet.rpcUrl,
                    contractAddress: config === null || config === void 0 ? void 0 : config.contractAddress,
                    providerName: wallet === null || wallet === void 0 ? void 0 : wallet.name,
                });
            },
        });
    }
}
exports.WebIrys = WebIrys;
exports.default = WebIrys;
