// @irys/sdk/build/cjs/web/tokens/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ethereum_1 = __importDefault(require("./ethereum"));
const near_1 = __importDefault(require("./near"));
const solana_1 = __importDefault(require("./solana"));
const erc20_1 = __importDefault(require("./erc20"));
const axios_1 = __importDefault(require("axios"));
const utils_1 = __importDefault(require("../../common/utils"));
const aptos_1 = __importDefault(require("./aptos"));
const ethersv5_1 = require("../providers/ethereum/ethersv5");
const ethersv6_1 = require("../providers/ethereum/ethersv6");
const arweave_1 = __importDefault(require("./arweave"));
function getTokenConfig({ irys, token, wallet, providerUrl, contractAddress, providerName, }) {
    switch (token) {
        case "ethereum":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "ethereum",
                ticker: "ETH",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://cloudflare-eth.com/",
                wallet: wallet,
            });
        case "matic":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "matic",
                ticker: "MATIC",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://polygon-rpc.com",
                wallet: wallet,
                minConfirm: 1,
            });
        case "arbitrum":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "arbitrum",
                ticker: "ETH",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://arb1.arbitrum.io/rpc",
                wallet: wallet,
            });
        case "bnb":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "bnb",
                ticker: "BNB",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://bsc-dataseed.binance.org",
                wallet: wallet,
            });
        case "avalanche":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "avalanche",
                ticker: "AVAX",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://api.avax.network/ext/bc/C/rpc",
                wallet: wallet,
            });
        case "boba-eth":
            return resolveProvider("ethereum", providerName, {
                irys: irys,
                name: "boba-eth",
                ticker: "ETH",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://mainnet.boba.network/",
                minConfirm: 1,
                wallet: wallet,
            });
        case "boba": {
            const k = new erc20_1.default({
                irys: irys,
                name: "boba",
                ticker: "BOBA",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://mainnet.boba.network/",
                contractAddress: contractAddress !== null && contractAddress !== void 0 ? contractAddress : "0xa18bF3994C0Cc6E3b63ac420308E5383f53120D7",
                minConfirm: 1,
                wallet: wallet,
            });
            // for L1 mainnet: "https://main-light.eth.linkpool.io/" and "0x42bbfa2e77757c645eeaad1655e0911a7553efbc"
            k.price = () => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const res = yield axios_1.default.post("https://api.livecoinwatch.com/coins/single", JSON.stringify({ currency: "USD", code: `${k.ticker}` }), {
                    headers: { "x-api-key": "75a7a824-6577-45e6-ad86-511d590c7cc8", "content-type": "application/json" },
                });
                yield utils_1.default.checkAndThrow(res, "Getting price data");
                if (!((_a = res === null || res === void 0 ? void 0 : res.data) === null || _a === void 0 ? void 0 : _a.rate)) {
                    throw new Error(`unable to get price for ${k.name}`);
                }
                return +res.data.rate;
            });
            return k;
        }
        case "solana":
            return new solana_1.default({
                irys: irys,
                name: "solana",
                ticker: "SOL",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://api.mainnet-beta.solana.com/",
                wallet: wallet,
            });
        // case "algorand":
        //     return new AlgorandConfig({ name: "algorand", ticker: "ALGO", providerUrl: providerUrl ?? "https://api.mainnet-beta.solana.com/", wallet: wallet })
        case "near":
            return new near_1.default({
                irys: irys,
                name: "near",
                ticker: "NEAR",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://rpc.mainnet.near.org",
                wallet: wallet,
            });
        case "aptos":
            return new aptos_1.default({
                irys: irys,
                name: "aptos",
                ticker: "APTOS",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://fullnode.mainnet.aptoslabs.com/v1",
                wallet: wallet,
            });
        case "arweave":
            return new arweave_1.default({
                irys: irys,
                name: "arweave",
                ticker: "AR",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://arweave.net",
                wallet: wallet,
            });
        case "base-eth":
            return new ethereum_1.default({
                irys: irys,
                name: "base-eth",
                ticker: "ETH",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://mainnet.base.org/",
                minConfirm: 2,
                wallet: wallet,
            });
        case "usdc-eth":
            return new erc20_1.default({
                irys: irys,
                name: "usdc-eth",
                ticker: "USDC",
                providerUrl: providerUrl !== null && providerUrl !== void 0 ? providerUrl : "https://cloudflare-eth.com/",
                contractAddress: contractAddress !== null && contractAddress !== void 0 ? contractAddress : "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
                wallet,
            });
        default:
            throw new Error(`Unknown/Unsupported token ${token}`);
    }
}
exports.default = getTokenConfig;
// @ts-expect-error todo fix
function resolveProvider(family, providerName, config) {
    switch (family) {
        case "ethereum":
            switch (providerName) {
                case "ethersv5":
                    return new ethersv5_1.EthereumEthersV5(config);
                case "ethersv6":
                    return new ethersv6_1.EthereumEthersV6(config);
                default:
                    return new ethereum_1.default(config);
            }
    }
}
